Source/ASLocking.h

```diff
@@ -0,0 +1,149 @@
+//
+//  ASLocking.h
+//  Texture
+//
+//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <Foundation/Foundation.h>
+#import <pthread/sched.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * An extension of NSLocking that supports -tryLock.
+ */
+@protocol ASLocking <NSLocking>
+
+/// Try to take lock without blocking. Returns whether the lock was taken.
+- (BOOL)tryLock;
+
+@end
+
+/**
+ * A set of locks acquired during ASLockSequence.
+ */
+typedef struct {
+  unsigned count;
+  CFTypeRef _Nullable locks[32];
```

Comment:
Do we do any checks against having more locks as 32? it seems if we call `ASUnlockSet` we could go over 32 to unlock if added more.

Reply:
You're right, done